tests/commands/test_store_client.py

```diff
@@ -0,0 +1,321 @@
+    fake_cookiejar.save()
+
+    auth_holder._load_credentials()
+
+    # check credentials
+    loaded_cookies = list(auth_holder._cookiejar)
+    assert len(loaded_cookies) == 1
+    assert loaded_cookies[0].value == fake_cookie.value  # compare the value as no __eq__ in Cookie
+    assert isinstance(auth_holder._cookiejar, MozillaCookieJar)
+    assert auth_holder._old_cookies == list(auth_holder._cookiejar)
+
+    # check other internal objects
+    assert isinstance(auth_holder._client, httpbakery.Client)
+    assert list(auth_holder._client.cookies)[0].value == fake_cookie.value
+    (im,) = auth_holder._client._interaction_methods
+    assert isinstance(im, httpbakery.WebBrowserInteractor)
+    assert im._open_web_browser == visit_page_with_browser
+
+
+def test_authholder_credentials_load_file_present_problem(auth_holder, caplog):
+    """Support the file to be corrupt (starts blank)."""
+    caplog.set_level(logging.WARNING, logger="charmcraft.commands")
+
+    with open(auth_holder._cookiejar_filepath, 'wb') as fh:
+        fh.write(b'this surely is not a valid cookie format :p')
+
+    auth_holder._load_credentials()
+
+    assert "Failed to read credentials" in caplog.records[0].message
+    assert auth_holder._old_cookies == []
+
+
+def test_authholder_credentials_load_file_missing(auth_holder, caplog):
+    """Support the file to not be there (starts blank)."""
+    caplog.set_level(logging.DEBUG, logger="charmcraft.commands")
+
+    auth_holder._load_credentials()
+
+    expected = "Credentials file not found: {!r}".format(auth_holder._cookiejar_filepath)
+    assert [expected] == [rec.message for rec in caplog.records]
+    assert auth_holder._old_cookies == []
+
+
+def test_authholder_credentials_save_notreally(auth_holder):
+    """Save does not really save if cookies didn't change."""
+    # create some fake cookies
+    fake_cookiejar = MozillaCookieJar(auth_holder._cookiejar_filepath)
+    fake_cookie = get_cookie()
+    fake_cookiejar.set_cookie(fake_cookie)
+    fake_cookiejar.save()
+
+    auth_holder._load_credentials()
+
+    with patch.object(auth_holder._cookiejar, 'save') as mock:
+        auth_holder._save_credentials()
+    assert mock.call_count == 0
+
+
+def test_authholder_credentials_save_reallysave(auth_holder):
+    """Save really do save if cookies changed."""
+    # create some fake cookies
+    fake_cookiejar = MozillaCookieJar(auth_holder._cookiejar_filepath)
+    fake_cookie = get_cookie()
+    fake_cookiejar.set_cookie(fake_cookie)
+    fake_cookiejar.save()
+
+    auth_holder._load_credentials()
+
+    with open(auth_holder._cookiejar_filepath, 'rb') as fh:
+        prv_file_content = fh.read()
+
+    other_cookie = get_cookie(value='different')
+    auth_holder._cookiejar.set_cookie(other_cookie)
+    auth_holder._save_credentials()
+
+    with patch.object(auth_holder._cookiejar, 'save') as mock:
+        auth_holder._save_credentials()
+    assert mock.call_count == 1
```

Comment:
I'm a bit confused by this test. As we call save multiple times but mock it out some of the time but then assert that the content is actually updated.

Reply:
It's called twice to check that the file was changed, and that it was calling the cookiejar's save. I refactored the test method to make it more clear, and added comments.